### PR TITLE
Fix pipes

### PR DESCRIPTION
**Pull request by erik-leven**
Added collapsible code blocks in best-practice as well as some re-writing in placed based on feedback from the consultants

### CODE REVIEW

File: best-practice.rst [hunk 1]
```diff
@@ -691,18 +741,118 @@ Below is a whole entity of the above global pipe and as seen, it gives an aggreg
...
+The aggregated data residing in a global dataset often needs to be transformed and/or enriched before it can be delivered to targets. Transforming and enriching data to ready it for delivery is implemented through preparation pipes. Preparation pipes use the aggregated entities from global datasets to combine and narrow the data down to what is necessary/required by the recipient system. The :ref:`filtering <filter_dtl_function>`  and relating of data are performed using the :ref:`RDF types <best_practice_rdf_type>` introduced earlier. Data can also be augmented performing hops to other datasets, for example a city-name can be fetched from a different dataset using the difi-postnummer. The goal is to have the data ready to be picked up by the outbound pipe.
+
+Below is an example of a preparation pipe, based on the global pipe above, where we wish to update the data in the system generating the hr-person data with crm-person data. More precicely, we wish to update data in the HR system to reflect the golden records from global-person. In addition we wish to update the HR data for "StreetAddress" with the address stated in the crm-person key "Address".
```
Comment (TarjeiSkrede) [inline best-practice.rst:754]: Typo: precicely

File: best-practice.rst [hunk 2]
```diff
@@ -691,18 +741,118 @@ Below is a whole entity of the above global pipe and as seen, it gives an aggreg
...
+            ["in", "~:crm:person", "_S.rdf:type"]
+          ],
+          ["copy",
+            ["list", "hr-person:SSN", "_id"]
+          ],
+          ["filter",
+            ["is-not-null", "_S.crm-person:Address"]
+          ],
+          ["add", "StreetAddress", "_S.crm-person:Address"],
+          ["add", "GivenName", "_S.global-person:firstname"],
```
Comment (TarjeiSkrede) [inline best-practice.rst:784]: Very minor comment: you use "GivenName" with a capital N here, but in "Surname" you don't.

File: best-practice.rst [hunk 3]
```diff
@@ -691,18 +741,118 @@ Below is a whole entity of the above global pipe and as seen, it gives an aggreg
...
+      "type": "dataset",
+      "dataset": "address-hr"
+    },
+    "sink": {
+      "type": "sql",
+      "system": "HR",
+      "primary_key": "SSN",
+      "table": "SomeTable"
+    },
+    "transform": {
```
Comment (TarjeiSkrede) [inline best-practice.rst:838]: Do we need to show a DTL transform here, as it merely copies all the properties of an entity and is essentially unnecessary (if we're thinking best practice, we should not add DTL transforms that only do this)?

File: configuration.rst [hunk 4]
```diff
@@ -915,14 +914,85 @@ The table below shows which strategy is chosen depending on the value of the pro
...
...rces). This "_updated" value is NOT the same as the "_updated" value seen in the output-tab in the Sesam GUI, but the value seen in the "pipe_offset" value in the execution log. In the two example codes below there's a pipe config and a corresponding code snippet from a microservice showing how continuation support can be set up for those cases. In this example, "supports_since" is activate in the pipe, which results in Sesam passing the "pipe_offset" to the microservice as a query parameter. The first time the pipe runs (or at a reset of the pipe) this value will be "None", resulting in the microservice requesting all the entities in the source. This specific source system has a property/column named "modifiedon". This source property contains a time-stamp value corresponding to the last time the entity was modified. This value is attached to a new entity property named "_updated". After the pipe has run Sesam will store the max value of all the entities "_updated" value as the new "pipe_offset". The next time the pipe runs this value will get passed to the microservice as the query parameter "since" such that the new request only covers data changed from after this specific time.
```
Comment (TarjeiSkrede) [inline configuration.rst:926]: I think you maybe should use another word than 'manually' or rewrite this sentence a little. It sounds a little bit like this is something you need to do manually, but should be more in terms that this is something the microservice needs to handle in its code.

File: configuration.rst [hunk 5]
```diff
@@ -915,14 +914,85 @@ The table below shows which strategy is chosen depending on the value of the pro
...
+
+In this case the data from the source is not ordered chronologically, which means we can not use the "is_chronological" tag. The benefit of chronologically ordered data in the source system is that should the pipe's pump for some reason fail in the middle of a request, Sesam can use the chronological order of the source data to continue requesting data from the last received entity. If the data is not ordered, Sesam has to re-run the whole last request.
```
Comment (TarjeiSkrede) [inline configuration.rst:928]: The benefit of chronologically ordered data in the source system is that if the pipe's pump for some reason should fail in the middle of a request...

File: configuration.rst [hunk 6]
```diff
@@ -915,14 +914,85 @@ The table below shows which strategy is chosen depending on the value of the pro
...
+          url = api_url + "/contacts"
+      else:
+          url = api_url + "/contacts?filter=modifiedon ge {}".format(request.args.get('since'))
+      headers = {"Authorization": "Bearer {}".format(token)}
+
+      req = requests.get(url = url, headers = headers)
+      if req.status_code != 200:
+        logger.error("Unexpected response status code: %d with response text %s" % (req.status_code, req.text))
+        raise AssertionError ("Unexpected response status code: %d with response text %s"%(req.status_code, req.text))
+      entities = req.json()["value"]
```
Comment (TarjeiSkrede) [inline configuration.rst:988]: Add a linebreak between "entities = req.json()["value"]" and "for entity in entities:" so it  becomes more clear what must be done in order to get continuation support to function in a microservice.

File: best-practice.rst [hunk 7]
```diff
@@ -317,7 +328,21 @@ Workflow for transforming data in Sesam
...
+
+
+**Inbound pipes** should preferably not change the source data, but rather tag the data for easier filtering further down in the Sesam data flow. The two standard ways of taging data in Sesam are through namespaced identifiers and RDF types. Namespaced identifiers are tags on column level which enables us to distinguish between properties from different sources with the same name. Namespaced identifiers are also often used to create the Sesam equivalent of foreign keys for easier join criteria further down the Sesam flow. RDF types are tags on row level which enables us to distinguish between entites from different sources in Sesam.
```
Comment (TarjeiSkrede) [inline best-practice.rst:345]: You could add that making references (ni's, kind of like sql joins) is also preferrable addedin the inbound pipes.

File: best-practice.rst [hunk 8]
```diff
@@ -317,7 +328,21 @@ Workflow for transforming data in Sesam
...
+|
+
+.. image:: images/best-practice/sesam-flow.png
+    :width: 800px
+    :align: center
+    :alt: Generic pipe concept  
+
+|
+
+The data from the source system is fed into Sesam through **inbound pipes** which collects and tags the data for further processing. During this step we do not with to change the data, but rather keep it as close to it's original state as possible. Depending on the semantical context of the data, it is sent into one of possibly several global pipes. We wish to store data in global pipes to better facilitate the consumption and improve the reusability of the data within Sesam. Connected to the global datasets generated by the global pipes are the preparation pipes. Preparation pipes are where the data is sorted, transformed and enriched to suit the requirements of the traget system. Finally the data is sent through an outbound pipe which is connected to a target system.
```
Comment (TarjeiSkrede) [inline best-practice.rst:342]: Typo: "we do not with to change the data".

File: best-practice.rst [hunk 9]
```diff
@@ -691,18 +744,110 @@ Below is a whole entity of the above global pipe and as seen, it gives an aggreg
...
+The aggregated data residing in a global dataset often needs to be transformed and/or enriched before it can be delivered to targets. Transforming and enriching data to ready it for delivery is implemented through preparation pipes. Preparation pipes use the aggregated entities from global datasets to combine and narrow the data down to what is necessary/required by the recipient system. The :ref:`filtering <filter_dtl_function>`  and relating of data are performed using the :ref:`RDF types <best_practice_rdf_type>` introduced earlier. Data can also be augmented performing hops to other datasets, for example a city-name can be fetched from a different dataset using the difi-postnummer. The goal is to have the data ready to be picked up by the outbound pipe.
```
Comment (TarjeiSkrede) [inline best-practice.rst:755]: One space too many in " The :ref:`filtering <filter_dtl_function>`  and".
"Data can also be augmented performing hops" should probably be "Data can also be augmented by performing hops".

File: best-practice.rst [hunk 10]
```diff
@@ -317,7 +328,21 @@ Workflow for transforming data in Sesam
...
+
+
+**Inbound pipes** should preferably not change the source data, but rather tag the data for easier filtering further down in the Sesam data flow. The two standard ways of taging data in Sesam are through namespaced identifiers and RDF types. Namespaced identifiers are tags on column level which enables us to distinguish between properties from different sources with the same name. Namespaced identifiers (in form of the :ref:`make-ni <dtl_transform-make-ni>` function) are also often used to create the Sesam equivalent of foreign keys for easier join criteria further down the Sesam flow. RDF types are tags on row level which enables us to distinguish between entites from different sources in Sesam.
```
Comment (TarjeiSkrede) [inline best-practice.rst:345]: Tiny typo: taging

File: best-practice.rst [hunk 11]
```diff
@@ -691,18 +744,110 @@ Below is a whole entity of the above global pipe and as seen, it gives an aggreg
...
+        ]
+      }
+    }
+  }
+
+.. raw:: html
+
+   </details>
+
+In this case we need to make sure that we do not overwrite existing "StreetAddress" values with potential null-values (altough we should be able to trust the master data from the CRM system, misakes do occur). We do not have to perform the same check for the properties from our golden records since in this case the hr-person data is part of the coalesce in the global-person pipe. Note that even though several of these properties is already the same in both the HR system and the CRM system, they do not have to be. The first time this integration runs there might be some unnecessary updates, but Sesam's built-in :ref:`change tracking <concepts-change-tracking>` will make sure only future changes will be passed through to the target system.
```
Comment (TarjeiSkrede) [inline best-practice.rst:798]: Tiny typo: misakes

File: best-practice.rst [hunk 12]
```diff
@@ -317,7 +328,21 @@ Workflow for transforming data in Sesam
...
+|
+
+.. image:: images/best-practice/sesam-flow.png
+    :width: 800px
+    :align: center
+    :alt: Generic pipe concept  
+
+|
+
+The data from the source system is fed into Sesam through **inbound pipes** which collects and tags the data for further processing. During this step we do not wish to change the data, but rather keep it as close to it's original state as possible. Depending on the semantical context of the data, it is sent into one of possibly several global pipes. We wish to store data in global pipes to better facilitate the consumption and improve the reusability of the data within Sesam. Connected to the global datasets generated by the global pipes are the preparation pipes. Preparation pipes are where the data is sorted, transformed and enriched to suit the requirements of the traget system. Finally the data is sent through an outbound pipe which is connected to a target system.
```
Comment (TarjeiSkrede) [inline best-practice.rst:342]: Another tiny typo: traget

Comment (TarjeiSkrede) [inline best-practice.rst:342]: Would maybe simplify the reading of this sentence: "Connected to the global datasets generated by the global pipes are the preparation pipes."

"Connected to the global datasets, which are generated by the global pipes, are the preparation pipes."

**Review by TarjeiSkrede — APPROVED** [review]

**Review by grove — APPROVED** [review]

**Pull request merged**